File: src/lib/api/expenses.ts
import { invokeCommand } from "../tauri";

export interface ExpenseResponse {
  id: string;
  category: string;
  description: string | null;
  amount: number;
  expense_date: string;
  payment_method: string | null;
  vendor: string | null;
  notes: string | null;
  is_deleted: boolean;
  deleted_at: string | null;
  created_at: string;
  updated_at: string;
}

export interface CreateExpenseRequest {
  category: string;
  amount: number;
  expense_date: string;
  description: string | null;
  payment_method: string | null;
  vendor: string | null;
  notes: string | null;
}

export interface UpdateExpenseRequest {
  category: string;
  amount: number;
  expense_date: string;
  description: string | null;
  payment_method: string | null;
  vendor: string | null;
  notes: string | null;
}

export async function createExpense(
  request: CreateExpenseRequest,
): Promise<ExpenseResponse> {
  return invokeCommand<ExpenseResponse>("create_expense", { request });
}

export async function getExpense(id: string): Promise<ExpenseResponse> {
  return invokeCommand<ExpenseResponse>("get_expense", { id });
}

export async function updateExpense(
  id: string,
  request: UpdateExpenseRequest,
): Promise<ExpenseResponse> {
  return invokeCommand<ExpenseResponse>("update_expense", { id, request });
}

export async function deleteExpense(id: string): Promise<void> {
  return invokeCommand<void>("delete_expense", { id });
}

export async function restoreExpense(id: string): Promise<ExpenseResponse> {
  return invokeCommand<ExpenseResponse>("restore_expense", { id });
}

export async function listExpenses(args: {
  search?: string;
  category?: string;
  date_from?: string;
  date_to?: string;
}): Promise<ExpenseResponse[]> {
  return invokeCommand<ExpenseResponse[]>("list_expenses", {
    search: args.search ?? null,
    category: args.category ?? null,
    dateFrom: args.date_from ?? null,
    dateTo: args.date_to ?? null,
  });
}

export async function totalExpenses(
  dateFrom: string,
  dateTo: string,
): Promise<number> {
  return invokeCommand<number>("total_expenses", {
    dateFrom: dateFrom,
    dateTo: dateTo,
  });
}

export const EXPENSE_CATEGORIES = [
  "Rent",
  "Electricity",
  "Water",
  "Gas",
  "Internet",
  "Equipment",
  "Maintenance",
  "Cleaning",
  "Supplies",
  "Staff",
  "Marketing",
  "Salary",
  "Other",
];

export const EXPENSE_PAYMENT_METHODS = [
  "Cash",
  "Card",
  "Bank Transfer",
  "Other",
];

File: src-tauri/src/commands/expenses.rs
use tauri::State;

use crate::database::Database;
use crate::dto::expense::{CreateExpenseRequest, ExpenseResponse, UpdateExpenseRequest};
use crate::errors::AppError;
use crate::services::expense_service;

use super::db::run_db;

#[tauri::command]
pub async fn create_expense(
    state: State<'_, Database>,
    request: CreateExpenseRequest,
) -> Result<ExpenseResponse, AppError> {
    let conn = state.inner().clone_conn();
    run_db(conn, move |c| expense_service::create_expense(c, request)).await
}

#[tauri::command]
pub async fn get_expense(
    state: State<'_, Database>,
    id: String,
) -> Result<ExpenseResponse, AppError> {
    let conn = state.inner().clone_conn();
    run_db(conn, move |c| expense_service::get_expense(c, &id)).await
}

#[tauri::command]
pub async fn update_expense(
    state: State<'_, Database>,
    id: String,
    request: UpdateExpenseRequest,
) -> Result<ExpenseResponse, AppError> {
    let conn = state.inner().clone_conn();
    run_db(conn, move |c| {
        expense_service::update_expense(c, &id, request)
    })
    .await
}

#[tauri::command]
pub async fn delete_expense(state: State<'_, Database>, id: String) -> Result<(), AppError> {
    let conn = state.inner().clone_conn();
    run_db(conn, move |c| expense_service::delete_expense(c, &id)).await
}

#[tauri::command]
pub async fn list_expenses(
    state: State<'_, Database>,
    search: Option<String>,
    category: Option<String>,
    date_from: Option<String>,
    date_to: Option<String>,
) -> Result<Vec<ExpenseResponse>, AppError> {
    let conn = state.inner().clone_conn();
    run_db(conn, move |c| {
        expense_service::list_expenses(
            c,
            search.as_deref().unwrap_or(""),
            category.as_deref(),
            date_from.as_deref(),
            date_to.as_deref(),
        )
    })
    .await
}

#[tauri::command]
pub async fn total_expenses(
    state: State<'_, Database>,
    date_from: String,
    date_to: String,
) -> Result<i64, AppError> {
    let conn = state.inner().clone_conn();
    run_db(conn, move |c| {
        expense_service::total_expenses(c, &date_from, &date_to)
    })
    .await
}

#[tauri::command]
pub async fn restore_expense(
    state: State<'_, Database>,
    id: String,
) -> Result<ExpenseResponse, AppError> {
    let conn = state.inner().clone_conn();
    run_db(conn, move |c| expense_service::restore_expense(c, &id)).await
}
